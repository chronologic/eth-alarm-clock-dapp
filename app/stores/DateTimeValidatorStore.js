import moment from 'moment';
import 'moment-timezone';

export default class DateTimeValidator {
  constructor() {
    var locale = window.navigator.userLanguage || window.navigator.language;
    moment.locale(locale);

    const localeData = moment.localeData();
    this.timeFormat = localeData.longDateFormat('LT');
    this.dateFormat = localeData.longDateFormat('L');
    this.dateTimeFormat = this.dateFormat + ' ' + this.timeFormat;
  }

  parse(date, time, tz) {
    return moment.tz(date + ' ' + time, this.dateTimeFormat, tz);
  }

  isValid(date, time, tz) {
    return this.parse(date, time, tz).isValid();
  }

  time(dateTime) {
    return moment(dateTime).format(this.timeFormat);
  }

  date(dateTime) {
    return moment(dateTime).format(this.dateFormat);
  }

  dateTime(dateTime) {
    return moment(dateTime).format(this.dateTimeFormat);
  }

  ts(date, time, tz) {
    return this.parse(date, time, tz).unix();
  }
}
