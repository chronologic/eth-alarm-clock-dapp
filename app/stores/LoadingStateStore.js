import { autorun, observable } from 'mobx';

class CORE_LOADING_STATES_NAMES {
  static START = 'START';
  static INITIALIZED_WEB3 = 'INITIALIZED_WEB3';
  static INITIALIZED_FEATURES = 'INITIALIZED_FEATURES';
  static INITIALIZED_TRANSACTION_STORE = 'INITIALIZED_TRANSACTION_STORE';
  static FETCHING_TRANSACTIONS = 'FETCHING_TRANSACTIONS';
  static FINISHING = 'FINISHING';
}

const CORE_LOADING_STATES_INFO = {
  [CORE_LOADING_STATES_NAMES.START]: {
    IS_COMPLETE: () => true
  },
  [CORE_LOADING_STATES_NAMES.INITIALIZED_WEB3]: {
    LOADING_MESSAGE: 'Connecting to blockchain node... Sometimes it can take up to a minute.',
    IS_COMPLETE: ({ _web3 }) => _web3.initialized
  },
  [CORE_LOADING_STATES_NAMES.INITIALIZED_FEATURES]: {
    LOADING_MESSAGE: 'Detecting supported features for the current environment...',
    IS_COMPLETE: ({ _features }) => _features.initialized
  },
  [CORE_LOADING_STATES_NAMES.INITIALIZED_TRANSACTION_STORE]: {
    LOADING_MESSAGE: 'Initializing service for fetching transactions...',
    IS_COMPLETE: ({ _transactionStore }) => _transactionStore.initialized
  },
  [CORE_LOADING_STATES_NAMES.FETCHING_TRANSACTIONS]: {
    IS_COMPLETE: ({ _transactionStore }) => !_transactionStore.info.gettingTransactions,
    LOADING_MESSAGE: 'Fetching transactions...'
  },
  [CORE_LOADING_STATES_NAMES.FINISHING]: {
    LOADING_MESSAGE: 'Fetching remaining data...',
    IS_COMPLETE: () => true
  }
};

export default class LoadingStateStore {
  @observable
  loadingState;

  loadingStates = [];
  loadingStatesInfo = {};

  _web3;

  constructor(web3Service, featuresService, transactionStore) {
    this._web3 = web3Service;
    this._features = featuresService;
    this._transactionStore = transactionStore;

    this.reset();

    autorun(() => this.refreshCurrentState());
  }

  get loadingMessage() {
    const nextLoadingStateIndex = this.loadingStates.indexOf(this.loadingState) + 1;
    const nextLoadingStateName = this.loadingStates[nextLoadingStateIndex];
    const nextLoadingState = this.loadingStatesInfo[nextLoadingStateName];

    if (nextLoadingState) {
      return nextLoadingState.LOADING_MESSAGE;
    }

    return this.loadingStatesInfo[CORE_LOADING_STATES_NAMES.FINISHING].LOADING_MESSAGE;
  }

  get progress() {
    const currentLoadingStateIndex = this.loadingStates.indexOf(this.loadingState);

    return Math.max(Math.ceil((currentLoadingStateIndex / this.loadingStates.length) * 100), 10);
  }

  add(name, info) {
    this.loadingStates.push(name);
    this.loadingStatesInfo[name] = info;
  }

  complete(name) {
    this.loadingStatesInfo[name].COMPLETED = true;
    this.refreshCurrentState();
  }

  reset() {
    this.loadingStates = Object.keys(CORE_LOADING_STATES_NAMES);
    this.loadingStatesInfo = Object.assign({}, CORE_LOADING_STATES_INFO);
    this.loadingState = CORE_LOADING_STATES_NAMES.START;
  }

  refreshCurrentState() {
    this.loadingStates.forEach((checkedLoadingStateName, checkedLoadingStateIndex) => {
      const currentLoadingStateIndex = this.loadingStates.indexOf(this.loadingState);

      if (
        currentLoadingStateIndex + 1 === checkedLoadingStateIndex &&
        this._checkLoadingStateCompleteness(checkedLoadingStateName)
      ) {
        this.loadingState = checkedLoadingStateName;
      }
    });
  }

  /**
   * @private
   *
   * @param {string} stateName
   */
  _checkLoadingStateCompleteness(stateName) {
    const state = this.loadingStatesInfo[stateName];

    return state.COMPLETED || (state.IS_COMPLETE && state.IS_COMPLETE(this));
  }
}
