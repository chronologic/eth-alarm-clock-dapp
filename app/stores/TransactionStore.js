import { showNotification } from '../services/notification';
import moment from 'moment';
import BigNumber from 'bignumber.js';
import { CONFIG } from '../lib/consts';

const requestFactoryStartBlocks = {
  3: 2594245,
  42: 5555500
};

const BLOCK_BUCKET_SIZE = 240;
const TIMESTAMP_BUCKET_SIZE = 3600;

export const DEFAULT_LIMIT = 10;

export class TRANSACTION_STATUS {
  static SCHEDULED = 'Scheduled';
  static EXECUTED = 'Executed';
  static FAILED = 'Failed';
  static CANCELLED = 'Cancelled';
  static MISSED = 'Not executed';
}

export class TEMPORAL_UNIT {
  static BLOCK = 1;
  static TIMESTAMP = 2;
}

const PARAMS_ERROR_TO_MESSAGE_MAPPING = {
  EmptyToAddress: 'Please enter recipient address.',
  CallGasTooHigh: 'Call gas is too high.',
  ExecutionWindowTooSoon: 'Execution window too soon. Please choose later date.',
  InsufficientEndowment:
    'Automatically calculated endowment is insufficient. Please contact developers.',
  ReservedWindowBiggerThanExecutionWindow: 'Reserved window is bigger than execution window.',
  InvalidTemporalUnit: 'Temporal unit is invalid. It should be either block or time.'
};

const SCHEDULING_GAS_LIMIT = 1500000;

export class TransactionStore {
  _eac;
  _web3;
  _fetcher;
  _eacScheduler;
  _cache;
  isSetup = false;

  _features;

  _requestFactory;

  constructor(eac, web3, fetcher, cache, featuresService) {
    this._web3 = web3;
    this._eac = eac;
    this._fetcher = fetcher;
    this._cache = cache;
    this._features = featuresService;

    this.setup();
  }

  // Returns an array of only the addresses of all transactions
  get allTransactionsAddresses() {
    return this._fetcher.allTransactionsAddresses;
  }

  get requestFactoryStartBlock() {
    const { network } = this._web3;
    return requestFactoryStartBlocks[network.id] || 0;
  }

  async setup() {
    if (this.isSetup) {
      return;
    }

    await this._web3.awaitInitialized();

    if (!this._features.isCurrentNetworkSupported) {
      return;
    }

    if (!this._requestFactory) {
      this._requestFactory = await this._eac.requestFactory();
    }

    this._eacScheduler = this._eacScheduler || (await this._eac.scheduler());

    this._fetcher.requestFactoryStartBlock = this.requestFactoryStartBlock;
    this._fetcher.startLazy();

    this.isSetup = true;
  }

  async awaitSetup() {
    if (this.isSetup) {
      return true;
    }

    return await new Promise(resolve => {
      setTimeout(async () => {
        resolve(await this.awaitSetup());
      }, 100);
    });
  }

  async getTransactions({ startBlock, endBlock = 'latest' }, cached) {
    await this.setup();

    startBlock = startBlock || this.requestFactoryStartBlock; // allow all components preload
    return await this._fetcher.getTransactions({ startBlock, endBlock }, cached);
  }

  async getAllTransactions(cached) {
    const transactions = await this._fetcher.getAllTransactions(cached);

    for (let transaction of transactions) {
      transaction.status = await this.getTxStatus(transaction);
    }

    return transactions;
  }

  async getAllTransactionAddresses() {
    if (
      this._fetcher.allTransactionsAddresses &&
      this._fetcher.allTransactionsAddresses.length > 0
    ) {
      return this._fetcher.allTransactionsAddresses;
    }

    return await this._fetcher.getTransactions({}, true, true);
  }

  async getBucketsForLastHours(hours) {
    const currentTimestamp = moment().unix();

    const buckets = [];

    let timestampBucket = await this.calcBucketForTimestamp(currentTimestamp);
    let blockBucket = await this.calcBucketForBlock(this.lastBlock);

    // Adding 0.5, because for each hour we fetch 2 buckets: timestamp, block.
    for (let i = 0; i < hours; i += 0.5) {
      // First, we fetch timestamp bucket, then block bucket.
      const isTimestampBucket = i % 1 === 0;

      buckets.push(isTimestampBucket ? timestampBucket : blockBucket);

      if (isTimestampBucket) {
        timestampBucket -= TIMESTAMP_BUCKET_SIZE;
      } else {
        /*
         * Since blockBucket is negative number we should add it to block bucket size,
         * if we want to go back in time.
         */
        blockBucket += BLOCK_BUCKET_SIZE;
      }
    }

    return buckets;
  }

  /**
   * @private
   * @returns Promise<{ transactions: Array }>
   */
  async _queryTransactions({
    transactions,
    offset,
    limit,
    pastHours,
    resolved,
    unresolved,
    sortByTimestampAscending
  }) {
    const processed = [];
    let total = 0;
    let buckets;

    if (pastHours) {
      buckets = await this.getBucketsForLastHours(pastHours);
    }

    for (const transaction of transactions) {
      const isResolved = await this.isTransactionResolved(transaction);
      let includeTransaction = false;

      if ((isResolved && resolved) || (!isResolved && unresolved)) {
        includeTransaction = true;
      }

      if (pastHours && includeTransaction) {
        includeTransaction = buckets.includes(transaction.getBucket());
      }

      if (includeTransaction) {
        processed.push(transaction);
      }
    }

    transactions = processed;

    if (sortByTimestampAscending) {
      const currentBlockTimestamp = await this._eac.Util.getTimestampForBlock(
        this._fetcher.lastBlock
      );

      transactions = transactions.sort((a, b) => {
        const aTimestamp = this.getTxTimestampEstimation(a, currentBlockTimestamp);
        const bTimestamp = this.getTxTimestampEstimation(b, currentBlockTimestamp);

        if (aTimestamp > bTimestamp) {
          return 1;
        }

        if (aTimestamp < bTimestamp) {
          return -1;
        }

        return 0;
      });
    }

    total = transactions.length;
    transactions = transactions.slice(offset, offset + limit);

    return {
      transactions,
      total
    };
  }

  // ------ UTILS ------
  async calcBucketForTimestamp(timestamp) {
    await this.awaitSetup();

    return this._requestFactory.calcBucket(timestamp, TEMPORAL_UNIT.TIMESTAMP);
  }

  async calcBucketForBlock(blockNumber) {
    await this.awaitSetup();

    return this._requestFactory.calcBucket(blockNumber, TEMPORAL_UNIT.BLOCK);
  }

  async getTransactionsFiltered({
    startBlock,
    endBlock,
    limit = DEFAULT_LIMIT,
    offset = 0,
    pastHours,
    resolved,
    unresolved,
    sortByTimestampAscending = true
  }) {
    let transactions = await this.getTransactions({ startBlock, endBlock });

    if (resolved || unresolved) {
      return this._queryTransactions({
        transactions,
        offset,
        limit,
        pastHours,
        resolved,
        unresolved,
        sortByTimestampAscending
      });
    }

    const total = transactions.length;

    transactions = transactions.slice(offset, offset + limit);

    return {
      transactions,
      total
    };
  }

  async getRequestsByOwner(ownerAddress, { limit = DEFAULT_LIMIT, offset = 0 }) {
    if (!this._requestFactory) {
      this._requestFactory = await this._eac.requestFactory();
    }

    const transactionsAddresses = await this._requestFactory.getRequestsByOwner(ownerAddress);
    let transactions = [];

    for (let address of transactionsAddresses) {
      const tx = await this._eac.transactionRequest(address);
      await tx.fillData();
      transactions.push(tx);
    }

    return {
      transactions: transactions.slice(offset, offset + limit),
      total: transactions.length
    };
  }

  getTxTimestampEstimation(transaction, currentBlockTimestamp) {
    const isTimestamp = this.isTxUnitTimestamp(transaction);

    const windowStart = transaction.windowStart.toNumber
      ? transaction.windowStart.toNumber()
      : transaction.windowStart;

    if (isTimestamp) {
      return windowStart;
    }

    let time;

    if (this._fetcher.lastBlock > windowStart) {
      time = windowStart;
    } else {
      const difference = windowStart - this._fetcher.lastBlock;

      time = currentBlockTimestamp + difference * CONFIG.averageBlockTime;
    }

    return time;
  }

  async getTxStatus(transaction) {
    let status = TRANSACTION_STATUS.SCHEDULED;

    if (transaction.wasCalled) {
      status = transaction.data.meta.wasSuccessful
        ? TRANSACTION_STATUS.EXECUTED
        : TRANSACTION_STATUS.FAILED;
    }

    if (transaction.isCancelled) {
      status = TRANSACTION_STATUS.CANCELLED;
    }

    if (await this.isTransactionMissed(transaction)) {
      status = TRANSACTION_STATUS.MISSED;
    }

    return status;
  }

  async getTransactionByAddress(address) {
    return await this._eac.transactionRequest(address, this._web3);
  }

  async isTransactionResolved(transaction) {
    const isMissed = await this.isTransactionMissed(transaction);

    return isMissed || transaction.wasCalled || transaction.isCancelled;
  }

  async isTransactionMissed(transaction) {
    let afterExecutionWindow;

    if (this.isTxUnitTimestamp(transaction)) {
      afterExecutionWindow = transaction.executionWindowEnd.lessThan(moment().unix());
    } else {
      afterExecutionWindow = transaction.executionWindowEnd.lessThan(this._fetcher.lastBlock);
    }

    return Boolean(afterExecutionWindow && !transaction.wasCalled);
  }

  async isTransactionFrozen(transaction) {
    return await transaction.inFreezePeriod();
  }

  isTxUnitTimestamp(transaction) {
    if (!transaction || !transaction.temporalUnit) {
      return false;
    }

    let temporalUnit = transaction.temporalUnit;

    if (transaction.temporalUnit.toNumber) {
      temporalUnit = transaction.temporalUnit.toNumber();
    }

    return temporalUnit === TEMPORAL_UNIT.TIMESTAMP;
  }

  async cancel(transaction, txParameters) {
    return await transaction.cancel(txParameters);
  }

  async refund(transaction, txParameters) {
    return await transaction.sendOwnerEther(txParameters);
  }

  async getBountiesForBucket(windowStart, isUsingTime) {
    let bucket;
    if (isUsingTime) {
      bucket = await this.calcBucketForTimestamp(windowStart);
    } else {
      bucket = await this.calcBucketForBlock(windowStart);
    }
    const transactions = await this._fetcher.getTransactionsInBuckets([bucket]);

    const { web3 } = this._web3;

    const bounties = [];
    let bounty, bountyInEth;

    transactions.forEach(tx => {
      bounty = tx.data.paymentData.bounty;
      bountyInEth = new BigNumber(web3.fromWei(bounty, 'ether'));
      bounties.push(bountyInEth);
    });

    return bounties;
  }

  async validateRequestParams(
    toAddress,
    callGas,
    callValue,
    windowSize,
    windowStart,
    gasPrice,
    fee,
    timeBounty,
    requiredDeposit,
    isTimestamp,
    endowment
  ) {
    if (!this._requestFactory) {
      this._requestFactory = await this._eac.requestFactory();
    }

    const temporalUnit = isTimestamp ? 2 : 1;
    const freezePeriod = isTimestamp ? 3 * 60 : 10; // 3 minutes or 10 blocks
    const reservedWindowSize = isTimestamp ? 5 * 60 : 16; // 5 minutes or 16 blocks
    const claimWindowSize = isTimestamp ? 60 * 60 : 255; // 60 minutes or 255 blocks
    const feeRecipient = '0x0'; // stub
    const fromAddress = this._web3.eth.defaultAccount;

    const serializedParams = [
      [fromAddress, feeRecipient, toAddress],
      [
        fee,
        timeBounty,
        claimWindowSize,
        freezePeriod,
        reservedWindowSize,
        temporalUnit,
        windowSize,
        windowStart,
        callGas,
        callValue,
        gasPrice,
        requiredDeposit
      ],
      endowment
    ];

    let paramsValid = false;
    let errors = [];

    try {
      const paramsValidBooleans = await this._requestFactory.validateRequestParams(
        ...serializedParams
      );

      errors = this._requestFactory.parseIsValid(paramsValidBooleans);

      paramsValid = errors.length === 0;
    } catch (error) {
      errors.push(error);
    }

    return {
      paramsValid,
      errors
    };
  }

  async schedule(
    toAddress,
    callData = '',
    callGas,
    callValue,
    windowSize,
    windowStart,
    gasPrice,
    fee,
    payment,
    requiredDeposit,
    waitForMined,
    isTimestamp
  ) {
    const endowment = this._eac.calcEndowment(callGas, callValue, gasPrice, fee, payment);

    const { paramsValid, errors } = await this.validateRequestParams(
      toAddress,
      callGas,
      callValue,
      windowSize,
      windowStart,
      gasPrice,
      fee,
      payment,
      requiredDeposit,
      isTimestamp,
      endowment
    );

    if (!paramsValid && errors.length > 0) {
      errors.forEach(
        error => error && showNotification(PARAMS_ERROR_TO_MESSAGE_MAPPING[error], 'danger', 4000)
      );

      return;
    }

    if (typeof this._eacScheduler === 'undefined') {
      this._eacScheduler = await this._eac.scheduler();
    }

    await this._eacScheduler.initSender({
      from: this._web3.eth.defaultAccount,
      gas: SCHEDULING_GAS_LIMIT,
      value: endowment
    });

    if (isTimestamp) {
      const receipt = await this._eacScheduler.timestampSchedule(
        toAddress,
        callData,
        callGas,
        callValue,
        windowSize,
        windowStart,
        gasPrice,
        fee,
        payment,
        requiredDeposit,
        waitForMined
      );

      return receipt;
    }

    const receipt = await this._eacScheduler.blockSchedule(
      toAddress,
      callData,
      callGas,
      callValue,
      windowSize,
      windowStart,
      gasPrice,
      fee,
      payment,
      requiredDeposit,
      waitForMined
    );

    return receipt;
  }
}
