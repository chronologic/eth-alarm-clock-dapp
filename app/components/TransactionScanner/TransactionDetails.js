import React from 'react';
import PropTypes from 'prop-types';
import { inject } from 'mobx-react';
import ScrollbarComponent from '../Common/ScrollbarComponent';
import { ValueDisplay } from '../Common/ValueDisplay';
import { BlockOrTimeDisplay } from './BlockOrTimeDisplay';
import { TRANSACTION_STATUS } from '../../stores/TransactionStore';
import { showNotification } from '../../services/notification';

const INITIAL_STATE = {
  callData: '',
  isTimestamp: false,
  status: '',
  transaction: {},
  executedAt: ''
};

@inject('transactionStore')
@inject('eacService')
@inject('web3Service')
class TransactionDetails extends ScrollbarComponent {
  state = INITIAL_STATE;

  _isMounted = false;

  constructor() {
    super(...arguments);

    this.state = INITIAL_STATE;
    this.cancelTransaction = this.cancelTransaction.bind(this);
  }

  getExecutedEvents(requestLib) {
    return new Promise(resolve => {
      requestLib.Executed({}, { fromBlock: 0, toBlock: 'latest' }).get((error, events) => {
        resolve(events);
      });
    });
  }

  async fetchData() {
    const { address, transactionStore } = this.props;

    const transaction = await transactionStore.getTransactionByAddress(address);

    await transaction.fillData();

    if (!this._isMounted) {
      return;
    }

    const requestLib = this.props.eacService.getRequestLibInstance(address);

    const events = await this.getExecutedEvents(requestLib);

    let executedAt = '';

    if (events[0]) {
      executedAt = events[0].transactionHash;
    }

    this.setState( {
      callData: await transaction.callData(),
      isTimestamp: transactionStore.isTxUnitTimestamp(transaction),
      status: await transactionStore.getTxStatus(transaction),
      transaction,
      executedAt,
      isFrozen: ''
    } );
  }

  async getFrozenStatus() {
    const { transactionStore } = this.props;
    const { transaction } = this.state;
    if (!transaction || !transaction.inFreezePeriod) {
      return;
    }
    const isFrozen = await transactionStore.isTransactionFrozen(transaction);
    this.setState( { isFrozen: isFrozen || transaction.isCancelled } );
  }

  async cancelTransaction() {
    const { transactionStore } = this.props;
    const { transaction } = this.state;

    const originalBodyCss = document.body.className;
    document.body.className += ' fade-me';
    this.cancelBtn.innerHTML = 'Canceling...';

    try {
      const success = await transactionStore.cancel(transaction);
      if (success) {
        this.setState({
          status: TRANSACTION_STATUS.CANCELLED
        });
      }
    } catch (error) {
      showNotification('Action cancelled by the user.', 'danger', 4000);
      this.cancelBtn.innerHTML = 'Cancel';
    }

    document.body.className = originalBodyCss;
  }

  async componentWillMount() {
    await this.fetchData();
  }

  isOwner( transaction){
    const { owner } = transaction;
    const { web3Service: { eth: { accounts } } } = this.props;
    const isOwner = accounts[0] == owner;
    return isOwner;
  }

  componentDidMount() {
    super.componentDidMount();
    this._isMounted = true;
    this.getFrozenStatus();
  }

  componentWillUnmount() {
    super.componentWillUnmount();
    this._isMounted = false;
  }

  getCancelSection() {
    const { transaction, isFrozen, status } = this.state;

    const isOwner = this.isOwner(transaction);

    if (isOwner && !isFrozen && status === TRANSACTION_STATUS.SCHEDULED) {
      return (
        <div className="text-center mt-5">
          <button className="btn btn-danger btn-cons"
            disabled={ isFrozen }
            onClick={ this.cancelTransaction }
            type="button"
            ref={(el) => this.cancelBtn = el}>
            <span>Cancel</span>
          </button>
          { isFrozen ? 'The transaction has been frozen.' : '' }
        </div>
      );
    }

    return <div></div>;
  }

  render() {
    const { callData, executedAt, isTimestamp, status, transaction } = this.state;
    const { bounty, callGas, callValue, fee, gasPrice, requiredDeposit, toAddress, windowStart, windowSize } = transaction;

    return (
      <div className="tab-pane slide active show">

        <table className="table">
          <tbody>
            <tr>
              <td>Status</td>
              <td>{status}<span className= { status !== TRANSACTION_STATUS.EXECUTED ? 'd-none' : '' } >&nbsp;at <a href={ this.props.web3Service.explorer + 'tx/' + executedAt }  target='_blank' rel='noopener noreferrer'>{ executedAt }</a></span></td>
            </tr>
            <tr>
              <td>To Address</td>
              <td><a href={this.props.web3Service.explorer + 'address/' + toAddress } target='_blank' rel='noopener noreferrer'>{ toAddress }</a></td>
            </tr>
            <tr>
              <td>Value/Amount</td>
              <td><ValueDisplay priceInWei= { callValue } /></td>
            </tr>
            <tr>
              <td>Data</td>
              <td>{callData}</td>
            </tr>
            <tr>
              <td>{ isTimestamp ? 'Time' : 'Block' }</td>
              <td><BlockOrTimeDisplay model= { windowStart } isTimestamp= { isTimestamp } /></td>
            </tr>
            <tr>
              <td>Window Size</td>
              <td><BlockOrTimeDisplay model= { windowSize } isTimestamp= { isTimestamp } duration= { true } /></td>
            </tr>
            <tr>
              <td>Gas Amount</td>
              <td> { callGas && callGas.toFixed() } </td>
            </tr>
            <tr>
              <td>Gas Price</td>
              <td><ValueDisplay priceInWei= { gasPrice } /></td>
            </tr>
            <tr>
              <td>Time Bounty</td>
              <td><ValueDisplay priceInWei= { bounty } /></td>
            </tr>
            <tr>
              <td>Fee</td>
              <td><ValueDisplay priceInWei= { fee } /></td>
            </tr>
            <tr>
              <td>Deposit</td>
              <td><ValueDisplay priceInWei= { requiredDeposit } /></td>
            </tr>
          </tbody>
        </table>

        {this.getCancelSection()}
      </div>
    );
  }
}

TransactionDetails.propTypes = {
  address: PropTypes.string,
  eacService: PropTypes.any,
  transactionStore: PropTypes.any
};

export default TransactionDetails;
