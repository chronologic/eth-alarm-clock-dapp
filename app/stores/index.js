import createBrowserHistory from 'history/createBrowserHistory';
import { RouterStore, syncHistoryWithStore } from 'mobx-react-router';
import { TransactionStore } from './TransactionStore';
import TimeNodeStore from './TimeNodeStore';
import { services } from '../services';
import ScheduleStore from './ScheduleStore';
import DateTimeValidatorStore from './DateTimeValidatorStore';
import { KeenStore } from './KeenStore';
import TransactionFetcher from './TransactionFetcher';
import TransactionCache from './TransactionCache';

const { eacService, featuresService, storageService, web3Service } = services;

const eacVersions = {
  client: require('@ethereum-alarm-clock/timenode-core').version,
  contracts: eacService.contracts,
  lib: eacService.version
};

const keenStore = new KeenStore(
  process.env.KEEN_PROJECT_ID,
  process.env.KEEN_WRITE_KEY,
  process.env.KEEN_READ_KEY,
  web3Service,
  eacVersions
);

const browserHistory = createBrowserHistory();
const routingStore = new RouterStore();
const scheduleStore = new ScheduleStore();
const dateTimeValidatorStore = new DateTimeValidatorStore();

export const transactionCache = new TransactionCache(storageService);
export const transactionFetcher = new TransactionFetcher(
  eacService,
  transactionCache,
  web3Service,
  featuresService
);
export const transactionStore = new TransactionStore(
  eacService,
  web3Service,
  transactionFetcher,
  transactionCache,
  featuresService
);
export const timeNodeStore = new TimeNodeStore(eacService, web3Service, keenStore, storageService);

export const history = syncHistoryWithStore(browserHistory, routingStore);

export const stores = {
  routing: routingStore,
  transactionStore,
  transactionCache,
  timeNodeStore,
  scheduleStore,
  dateTimeValidatorStore,
  keenStore
};
