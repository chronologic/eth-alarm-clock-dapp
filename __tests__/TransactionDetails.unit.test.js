import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'mobx-react';
import Web3Service from '../app/services/web3';
import BigNumber from 'bignumber.js';
import TransactionDetails from '../app/components/TransactionScanner/TransactionDetails';
import { TransactionStore } from '../app/stores/TransactionStore';
import TokenHelper from '../app/services/token-helper';
import LoadingStateStore from '../app/stores/LoadingStateStore';
import TransactionHelper from '../app/services/transaction-helper';

describe('TransactionDetails', () => {
  it('correctly renders', async () => {
    const TEST_EXPLORER = 'https://etherscan.io/';

    const web3Service = new Web3Service({
      eth: {
        accounts: []
      },
      explorer: TEST_EXPLORER,
      web3: {
        eth: {
          contract() {
            return {
              at() {
                return {
                  name() {}
                };
              }
            };
          }
        },
        sha3() {
          return '';
        }
      },
      connect() {}
    });

    const eacService = {
      getRequestLibInstance() {
        return {
          Executed() {
            return {
              get(callback) {
                callback(null, []);
              }
            };
          }
        };
      },

      scheduler() {
        return {};
      }
    };

    const transactionHelper = new TransactionHelper();

    const transactionStore = new TransactionStore(
      eacService,
      web3Service,
      {},
      {},
      {},
      transactionHelper
    );
    const tokenHelper = new TokenHelper(web3Service);

    const loadingStateStore = new LoadingStateStore(web3Service, {}, transactionStore);

    const injectables = {
      eacService,
      loadingStateStore,
      tokenHelper,
      transactionStore,
      web3Service
    };

    const transaction = {
      bounty: new BigNumber('10000'),
      callData() {
        return 'xyz';
      },
      toAddress: '0x123306090abab3a6e1400e9345bc60c78a8bef57',
      data: {
        wasSuccessful: true
      },
      afterExecutionWindow: () => false
    };

    let mockedRender = renderer.create(
      <Provider {...injectables}>
        <TransactionDetails transaction={transaction} />
      </Provider>
    );

    let tree = mockedRender.toJSON();
    expect(tree).toMatchSnapshot();

    const tableRows = tree.children.find(el => el.props.className === 'table d-block').children[0]
      .children;

    // `To Address` field checks
    const toAddressTitleDisplay = tableRows[1].children[0].children[0];

    expect(toAddressTitleDisplay).toEqual('To Address');

    const toAddressExplorerLink = tableRows[1].children[1].children[0];

    expect(toAddressExplorerLink.props.href).toEqual(
      `${TEST_EXPLORER}address/${transaction.toAddress}`
    );

    const toAddressValueDisplay = tableRows[1].children[1].children[0].children[0];

    expect(toAddressValueDisplay).toEqual(transaction.toAddress);
  });
});
