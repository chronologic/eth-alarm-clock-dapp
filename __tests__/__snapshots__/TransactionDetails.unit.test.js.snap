// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionDetails correctly renders 1`] = `
<div
  className="tab-pane slide active show"
>
  <div />
  <table
    className="table d-block"
  >
    <tbody
      className="d-block"
    >
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Status
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        >
          
          <span
            className="d-none"
          >
            Â at
             
            <a
              href="https://etherscan.io/tx/"
              rel="noopener noreferrer"
              target="_blank"
            >
              
            </a>
          </span>
        </td>
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          To Address
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        >
          <a
            href="https://etherscan.io/address/0x123306090abab3a6e1400e9345bc60c78a8bef57"
            rel="noopener noreferrer"
            target="_blank"
          >
            0x123306090abab3a6e1400e9345bc60c78a8bef57
          </a>
        </td>
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Value/Amount
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        />
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Data
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
          title=""
        >
          
        </td>
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Block
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        >
          
        </td>
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Window Size
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        >
          
        </td>
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Gas Amount
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        >
           
           
        </td>
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Gas Price
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        />
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Time Bounty
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        >
          10000 WEI
        </td>
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Fee
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        />
      </tr>
      <tr
        className="row"
      >
        <td
          className="d-inline-block col-5 col-md-3"
        >
          Deposit
        </td>
        <td
          className="d-inline-block col-7 col-md-9"
        />
      </tr>
    </tbody>
  </table>
  <div
    className="row"
  >
    <div
      className="col-2 col-sm-4 col-md-6 col-lg-8"
    />
    <div
      className="col-10 col-sm-8 col-md-6 col-lg-4"
    >
      <div
        className="row"
      >
        <div
          className="col-6"
        />
        <div
          className="col-6"
        />
      </div>
    </div>
    <div
      className="col-12"
    >
      <div />
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <div
        className="alert alert-info"
      >
        Note that transaction can be cancelled:
        <br />
        <ol
          className="list-normalized"
        >
          <li>
            Before 
            <b>
              
            </b>
             
            (claiming window start)
          </li>
          <li>
            When wasn't executed by any TimeNode after
             
            <b>
              
            </b>
             
            (execution window end)
          </li>
          <li>
            When it hasn't been already cancelled
          </li>
          <li>
            Only by its owner
          </li>
        </ol>
        <div
          className="mt-3"
        >
          <button
            className="btn btn-danger btn-cons"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <span>
              Cancel
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
